parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +622,1024 @@ protected GroupType build(String name) {
+        return listValue(Type.Repetition.OPTIONAL);
+      }
+    }
+
+    public static class MapGroupValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseGroupBuilder<P, MapGroupValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapGroupValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapMapValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseMapBuilder<P, MapMapValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapMapValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapMapValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListValueBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseListBuilder<P, MapListValueBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapListValueBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapListValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    public static class MapListKeyBuilder<P, Q extends BaseMapBuilder<P, Q>> extends
+        BaseListBuilder<P, MapListKeyBuilder<P, Q>> {
+      private final BaseMapBuilder<P, Q> parent;
+
+      public MapListKeyBuilder(BaseMapBuilder<P, Q> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        return super.named(name);
+      }
+
+      @Override
+      protected MapListKeyBuilder<P, Q> self() {
+        return this;
+      }
+
+      public void setKey(Type key) {
+        parent.setKeyType(key);
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public BaseMapBuilder(P parent) {
+      super(parent);
+    }
+
+    private BaseMapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected abstract Q self();
+
+    public MapKeyBuilder<P, Q> key(PrimitiveTypeName type) {
```

Comment:
I thought we had decided not to add a key builder for maps? I thought that was to be handled by `key(Type)` and `requiredKey(Type)` methods.